expression/builtin_cast.go

```diff
@@ -1983,3 +1986,87 @@ func WrapWithCastAsJSON(ctx sessionctx.Context, expr Expression) Expression {
 	}
 	return BuildCastFunction(ctx, expr, tp)
 }
+
+// TryPushCastDownToControlFunctionForHybridType try to push cast down to control function for Hybrid Type.
+// If necessary, it will rebuild control function using changed args.
+// When a hybrid type is the output of a control function, the result may be as a numeric type to subsequent calculation
+// We should perform the `Cast` operation early to avoid using the wrong type for calculation
+func TryPushCastDownToControlFunctionForHybridType(ctx sessionctx.Context, expr Expression, tp *types.FieldType) (res Expression, success bool) {
+	sf, ok := expr.(*ScalarFunction)
+	if !ok {
+		return expr, false
+	}
+
+	var wrapCastFunc func(ctx sessionctx.Context, expr Expression) Expression
+	switch tp.EvalType() {
+	case types.ETInt:
+		wrapCastFunc = WrapWithCastAsInt
+	case types.ETReal:
+		wrapCastFunc = WrapWithCastAsReal
+	case types.ETString:
+		wrapCastFunc = WrapWithCastAsString
+	default:
+		return expr, false
+	}
+
+	args := sf.GetArgs()
+	switch sf.FuncName.L {
+	case ast.If:
+		if args[1].GetType().Hybrid() || args[2].GetType().Hybrid() {
+			args[1] = wrapCastFunc(ctx, args[1])
+			args[2] = wrapCastFunc(ctx, args[2])
+			f, err := funcs[ast.If].getFunction(ctx, args)
+			if err != nil {
+				return expr, false
+			}
+			sf.RetType, sf.Function = f.getRetTp(), f
```

Comment:
Why not `return f, true `?

Reply:
`f` is a builtinFunc, it is a field in `ScalarFunction`